spec/rubocop/cop/badge_spec.rb

```diff
@@ -11,34 +11,59 @@
     expect(badge.cop_name).to eql('ModuleMustBeAClassCop')
   end
 
-  describe '.parse' do
-    it 'parses Department/CopName syntax' do
-      expect(described_class.parse('Foo/Bar'))
-        .to eq(described_class.new('Foo', 'Bar'))
+  describe '.new' do
+    shared_examples 'assignment of department and name' do |class_name_parts, department, name|
+      it 'assigns department' do
+        expect(described_class.new(class_name_parts).department).to eq(department)
+      end
+
+      it 'assigns name' do
+        expect(described_class.new(class_name_parts).cop_name).to eq(name)
+      end
     end
 
-    it 'parses unqualified badge references' do
-      expect(described_class.parse('Bar'))
-        .to eql(described_class.new(nil, 'Bar'))
+    include_examples 'assignment of department and name', %w[Foo], nil, 'Foo'
+    include_examples 'assignment of department and name', %w[Foo Bar], :Foo, 'Bar'
+    include_examples 'assignment of department and name', %w[Foo Bar Baz], :'Foo/Bar', 'Baz'
+    include_examples 'assignment of department and name', %w[Foo Bar Baz Qux], :'Foo/Bar/Baz', 'Qux'
+  end
+
+  describe '.parse' do
+    shared_examples 'cop identifier parsing' do |identifier, class_name_parts|
+      it 'parses identifier' do
+        expect(described_class.parse(identifier)).to eq(described_class.new(class_name_parts))
+      end
     end
+
+    include_examples 'cop identifier parsing', 'Bar', %w[Bar]
+    include_examples 'cop identifier parsing', 'Foo/Bar', %w[Foo Bar]
+    include_examples 'cop identifier parsing', 'Foo/Bar/Baz', %w[Foo Bar Baz]
+    include_examples 'cop identifier parsing', 'Foo/Bar/Baz/Qux', %w[Foo Bar Baz Qux]
   end
 
   describe '.for' do
-    it 'parses cop class name' do
-      expect(described_class.for('RuboCop::Cop::Foo::Bar'))
-        .to eq(described_class.new('Foo', 'Bar'))
+    shared_examples 'cop class name parsing' do |class_name, class_name_parts|
+      it 'parses cop class name' do
+        expect(described_class.for(class_name)).to eq(described_class.new(class_name_parts))
+      end
     end
+
+    include_examples 'cop class name parsing', 'Foo', %w[Foo]
+    include_examples 'cop class name parsing', 'Foo::Bar', %w[Foo Bar]
+    include_examples 'cop class name parsing', 'RuboCop::Cop::Foo', %w[Cop Foo]
```

Comment:
I think this line is wrong. We don’t want the class name `"RuboCop::Cop::Foo"` to be recognized as a cop `Foo` with department `Cop`. Since the string starts with `RuboCop::Cop`, `Foo` must clearly be the department name, and the cop’s name is missing. I would expect` Badge.for` to raise an error.

Reply:
Or, the department should be detected as none, and the cop name as `Foo`.

Reply:
Well, I think this is exactly how it was before. Previously method `for` was `new(*class_name.split('::').last(2))` so it would parse `RuboCop::Cop::Foo` as `Cop/Foo` (`Cop` department).

If I am to add something like this to `for` method
```
if class_name.start_with?('RuboCop::Cop', 'RuboCop::Base') && parts.length == 3
  raise
end
```
it would break a lot of specs. For example, we have spec/rubocop/cop/cop_spec.rb and in that file we have the following:
```
RSpec.describe RuboCop::Cop::Cop, :config do
  context 'with no submodule' do
    it('has right name') { expect(cop_class.cop_name).to eq('Cop/Cop') }
    it('has right department') { expect(cop_class.department).to eq(:Cop) }
  end
end
```

So... it seems it was intentional it doesn't raise an error? 

So I can add whatever thing you like here but maybe let's confirm it with the other core developers first since it would be a big change?

Reply:
I guess this was necessary during the period in which we had to support cops without a department (originally there were none and all cops were in the same namespace so to speak). I agree this should be changed.